src/audio/pipeline/pipeline-graph.c

```diff
@@ -423,29 +423,35 @@ int pipeline_for_each_comp(struct comp_dev *current,
 }
 
 /* visit connected pipeline to find the dai comp */
-struct comp_dev *pipeline_get_dai_comp(uint32_t pipeline_id)
+struct comp_dev *pipeline_get_dai_comp(uint32_t pipeline_id, int dir)
 {
-	struct ipc_comp_dev *sink;
+	struct ipc_comp_dev *crt;
 	struct ipc *ipc = ipc_get();
 
-	sink = ipc_get_ppl_sink_comp(ipc, pipeline_id);
-	while (sink) {
+	crt = ipc_get_ppl_comp(ipc, pipeline_id, dir);
+	while (crt) {
 		struct comp_buffer *buffer;
 		struct comp_dev *comp;
 
-		/* dai is found */
-		if (list_is_empty(&sink->cd->bsink_list))
-			return sink->cd;
+		/* based on given direction check if list of
+		 * sink or source buffers is empty. If so
+		 * then we have found the DAI.
+		 */
+		if ((dir == PPL_DIR_DOWNSTREAM &&
+		     list_is_empty(&crt->cd->bsink_list)) ||
+		    (dir == PPL_DIR_UPSTREAM &&
+		     list_is_empty(&crt->cd->bsource_list)))
```

Comment:
we have `comp_buffer_list()` for this, so this would be `list_is_empty(comp_buffer_list(crt->cd, dir))` or in fact save the result of `comp_buffer_list(crt->cd, dir)` and use it here and below in line 446

Reply:
Yes, this looks ok.